### PR TITLE
add readonly to beego middleware

### PR DESCRIPTION
**Pull request by wy65701436**
Signed-off-by: wang yan <wangyan@vmware.com>

### CODE REVIEW

File: src/server/middleware/readonly/readonly.go [hunk 1]
```diff
@@ -67,9 +67,11 @@ func MiddlewareWithConfig(config Config, skippers ...middleware.Skipper) func(ht
...
 	return middleware.New(func(w http.ResponseWriter, r *http.Request, next http.Handler) {
 		if config.ReadOnly(r) {
-			pkgE := internal_errors.New(nil).WithCode(internal_errors.DENIED).WithMessage("The system is in read only mode. Any modification is prohibited.")
-			serror.SendError(w, pkgE)
-			return
+			if r.Method == http.MethodDelete || r.Method == http.MethodPost || r.Method == http.MethodPatch || r.Method == http.MethodPut {
+				pkgE := internal_errors.New(nil).WithCode(internal_errors.DENIED).WithMessage("The system is in read only mode. Any modification is prohibited.")
+				serror.SendError(w, pkgE)
+				return
+			}
```
Comment (heww) [inline src/server/middleware/readonly/readonly.go]: These changes are not needed here, only need to join the custom's skipers and the safeMethodSkipper .

Comment (wy65701436) [reply src/server/middleware/readonly/readonly.go]: thanks, updated.

File: src/server/registry/route.go [hunk 2]
```diff
@@ -33,7 +33,8 @@ func RegisterRoutes() {
 	root := router.NewRoute().
 		Path("/v2").
 		Middleware(artifactinfo.Middleware()).
-		Middleware(v2auth.Middleware())
+		Middleware(v2auth.Middleware()).
+		Middleware(readonly.Middleware())
```
Comment (ywk253100) [inline src/server/registry/route.go]: The `/v2/xxx` requests should already be covered by the global middleware?

File: src/core/middlewares/middlewares.go [hunk 3]
```diff
@@ -29,11 +30,28 @@ import (
...
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/c/userExists")),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/c/oidc/onboard")),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/service/notifications/jobs/adminjob/"+numericRegexp.String())),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/service/notifications/jobs/replication/"+numericRegexp.String())),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/service/notifications/jobs/replication/task/"+numericRegexp.String())),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/service/notifications/jobs/webhook/"+numericRegexp.String())),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/service/notifications/jobs/retention/task/"+numericRegexp.String())),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/service/notifications/jobs/schedules/"+numericRegexp.String())),
+		middleware.MethodAndPathSkipper(http.MethodPost, match("^/service/notifications/jobs/webhook/"+numericRegexp.String())),
+	}
```
Comment (ywk253100) [inline src/core/middlewares/middlewares.go]: How about just cover the requests whose path has prefix of `/v2/` or `/api/` ? Just like what the `legacyAPISkipper` does

Seems the `legacyAPISkipper` should be renamed...

File: src/server/registry/route.go (hunk 2 above)
Comment (wy65701436) [reply src/server/registry/route.go]: yes, will remove it.

**Pull request merged**